<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>nmf_models.nmf_models_mod_updates module &mdash; sc_intNMF 0.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/graphviz.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.6.0/jquery.min.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="nmf_models.plotting module" href="nmf_models.plotting.html" />
    <link rel="prev" title="nmf_models package" href="nmf_models.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            sc_intNMF
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="nmf_models.html">nmf_models package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="nmf_models.html#submodules">Submodules</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">nmf_models.nmf_models_mod_updates module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#nmf_models.nmf_models_mod_updates.InvalidMatType"><code class="docutils literal notranslate"><span class="pre">InvalidMatType</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#nmf_models.nmf_models_mod_updates.intNMF"><code class="docutils literal notranslate"><span class="pre">intNMF</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#nmf_models.nmf_models_mod_updates.log_tf_idf"><code class="docutils literal notranslate"><span class="pre">log_tf_idf()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="nmf_models.plotting.html">nmf_models.plotting module</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="nmf_models.html#module-nmf_models">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">nmf_models</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="nmf_models.html">nmf_models package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="nmf_models.html#submodules">Submodules</a><ul class="current">
<li class="toctree-l4 current"><a class="current reference internal" href="#">nmf_models.nmf_models_mod_updates module</a></li>
<li class="toctree-l4"><a class="reference internal" href="nmf_models.plotting.html">nmf_models.plotting module</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="nmf_models.html#module-nmf_models">Module contents</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="readme.html">README</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">sc_intNMF</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="nmf_models.html">nmf_models package</a></li>
      <li class="breadcrumb-item active">nmf_models.nmf_models_mod_updates module</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="module-nmf_models.nmf_models_mod_updates">
<span id="nmf-models-nmf-models-mod-updates-module"></span><h1>nmf_models.nmf_models_mod_updates module<a class="headerlink" href="#module-nmf_models.nmf_models_mod_updates" title="Permalink to this heading"></a></h1>
<dl class="py exception">
<dt class="sig sig-object py" id="nmf_models.nmf_models_mod_updates.InvalidMatType">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">nmf_models.nmf_models_mod_updates.</span></span><span class="sig-name descname"><span class="pre">InvalidMatType</span></span><a class="reference internal" href="_modules/nmf_models/nmf_models_mod_updates.html#InvalidMatType"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nmf_models.nmf_models_mod_updates.InvalidMatType" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Exception</span></code></p>
<p>Raised if the mat is invalid.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="nmf_models.nmf_models_mod_updates.intNMF">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">nmf_models.nmf_models_mod_updates.</span></span><span class="sig-name descname"><span class="pre">intNMF</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_topics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epochs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">init</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'random'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod1_skew</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">l1_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nmf_models/nmf_models_mod_updates.html#intNMF"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nmf_models.nmf_models_mod_updates.intNMF" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class to run int NMF on multiome data</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="nmf_models.nmf_models_mod_updates.intNMF.sdasd">
<span class="sig-name descname"><span class="pre">sdasd</span></span><a class="headerlink" href="#nmf_models.nmf_models_mod_updates.intNMF.sdasd" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>(</strong><strong>k</strong><strong>)</strong> (<em>n_topics</em>) – </p></li>
<li><p><strong>epochs</strong> (<em>Number</em><em> of </em><em>interations during optimisation</em>) – </p></li>
<li><p><strong>init</strong> (<em>initialisation  method. random</em><em> | </em><em>svd</em>) – </p></li>
<li><p><strong>metrics.</strong> (<em>Lists to store training</em>) – </p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="nmf_models.nmf_models_mod_updates.intNMF._HALS">
<span class="sig-name descname"><span class="pre">_HALS</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">H</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">WtW</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">WtM</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eit1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nmf_models/nmf_models_mod_updates.html#intNMF._HALS"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nmf_models.nmf_models_mod_updates.intNMF._HALS" title="Permalink to this definition"></a></dt>
<dd><p>Optimizing min_{V &gt;= 0} ||M-UV||_F^2 with an exact block-coordinate descent.</p>
<p>UtU and UtM are exprensive to compute so multiple updates are calcuated.
stop condition
(epoch run time ) &lt; (precompute time + current iteration time)/2
AND
sum of squares of updates to V at current epoch &gt; 0.01 * sum of squares of updates to V at first epoch</p>
<section id="parameters">
<h2>Parameters:<a class="headerlink" href="#parameters" title="Permalink to this heading"></a></h2>
<dl class="simple">
<dt>V: Array-like</dt><dd><p>mat to update</p>
</dd>
<dt>UtU<span class="classifier">Array-like</span></dt><dd><p>precomputed dense array</p>
</dd>
<dt>UtM<span class="classifier">Array-like</span></dt><dd><p>precomputed dense array</p>
</dd>
<dt>eit1<span class="classifier">int</span></dt><dd><p>precompute time</p>
</dd>
<dt>alpha<span class="classifier">float</span></dt><dd><p>control time based stop criteria</p>
</dd>
<dt>delta<span class="classifier">float</span></dt><dd><p>control loss based stop criteria</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>V</strong> (<em>Array-like</em>) – array-like optimised array</p></li>
<li><p><strong>n_it</strong> (<em>int</em>) – number of iterations (usually 6/7)</p></li>
</ul>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="nmf_models.nmf_models_mod_updates.intNMF._HALS_W">
<span class="sig-name descname"><span class="pre">_HALS_W</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">W</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rnaHHt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rnaMHt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">atacHHt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">atacMHt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eit1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nmf_models/nmf_models_mod_updates.html#intNMF._HALS_W"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nmf_models.nmf_models_mod_updates.intNMF._HALS_W" title="Permalink to this definition"></a></dt>
<dd><p>Optimizing min_{W &gt;= 0} ||X1-WH1||_F^2 + ||X2-WH2||_F^2.</p>
<p>An exact block-coordinate descent scheme is applied to update W. HHt
and HtM (X) are exprensive to compute so multiple updates are
pre-calcuated.</p>
<section id="id1">
<h2>Parameters:<a class="headerlink" href="#id1" title="Permalink to this heading"></a></h2>
<dl class="simple">
<dt>W<span class="classifier">array-like</span></dt><dd><p>mat to update</p>
</dd>
<dt>atacHHt<span class="classifier">array-like</span></dt><dd><p>precomputed dense array</p>
</dd>
<dt>atacMHt<span class="classifier">array-like</span></dt><dd><p>precomputed dense array</p>
</dd>
<dt>rnaHHt<span class="classifier">array-like</span></dt><dd><p>precomputed dense array</p>
</dd>
<dt>rnaMHt<span class="classifier">array-like</span></dt><dd><p>precomputed dense array</p>
</dd>
<dt>eit1<span class="classifier">int</span></dt><dd><p>precompute time</p>
</dd>
<dt>alpha<span class="classifier">float</span></dt><dd><p>Control time based stop criteria</p>
</dd>
<dt>delta<span class="classifier">float</span></dt><dd><p>control loss based stop criteria</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>V</strong> (<em>optimised array</em>)</p></li>
<li><p><strong>n_it</strong> (<em>number of iterations (usually 6/7)</em>)</p></li>
<li><p><em>stop condition</em></p></li>
<li><p><em>(epoch run time ) &lt; (precompute time + current iteration time)/2</em></p></li>
<li><p><em>AND</em></p></li>
<li><p><em>sum of squares of updates to V at current epoch &gt; 0.01 * sum of squares of updates to V at first epoch</em></p></li>
</ul>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="nmf_models.nmf_models_mod_updates.intNMF._initialize_nmf">
<span class="sig-name descname"><span class="pre">_initialize_nmf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ATAC_mat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_components</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-06</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">init</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'random'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nmf_models/nmf_models_mod_updates.html#intNMF._initialize_nmf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nmf_models.nmf_models_mod_updates.intNMF._initialize_nmf" title="Permalink to this definition"></a></dt>
<dd><p>Algorithms for NMF initialization.</p>
<p>Computes an initial guess for the non-negative
rank k matrix approximation for X: X = WH</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>array-like</em><em>, </em><em>shape</em><em> (</em><em>n_samples</em><em>, </em><em>n_features</em><em>)</em>) – The data matrix to be decomposed.</p></li>
<li><p><strong>n_components</strong> (<em>integer</em>) – The number of components desired in the approximation.</p></li>
<li><p><strong>init</strong> (<em>None</em><em> | </em><em>'random'</em><em> | </em><em>'nndsvd'</em><em> | </em><em>'nndsvda'</em><em> | </em><em>'nndsvdar'</em>) – <p>Method used to initialize the procedure.
Default: None.
Valid options:
- None: ‘nndsvd’ if n_components &lt;= min(n_samples, n_features),</p>
<blockquote>
<div><p>otherwise ‘random’.</p>
</div></blockquote>
<ul>
<li><dl class="simple">
<dt>’random’: non-negative random matrices, scaled with:</dt><dd><p>sqrt(X.mean() / n_components)</p>
</dd>
</dl>
</li>
<li><p>’nndsvd’: Nonnegative Double Singular Value Decomposition (NNDSVD)</p></li>
</ul>
</p></li>
<li><p><strong>eps</strong> (<em>float</em>) – Truncate all values less then this in output to zero.</p></li>
<li><p><strong>random_state</strong> (<em>int</em><em>, </em><em>RandomState instance</em><em> or </em><em>None</em><em>, </em><em>optional</em><em>, </em><em>default: None</em>) – If int, random_state is the seed used by the random number generator;
If RandomState instance, random_state is the random number generator;
If None, the random number generator is the RandomState instance used
by <cite>np.random</cite>. Used when <code class="docutils literal notranslate"><span class="pre">random</span></code> == ‘nndsvdar’ or ‘random’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>W</strong> (<em>array-like, shape (n_samples, n_components)</em>) – Initial guesses for solving X ~= WH</p></li>
<li><p><strong>H</strong> (<em>array-like, shape (n_components, n_features)</em>) – Initial guesses for solving X ~= WH</p></li>
<li><p><strong>H_atac</strong> (<em>array-like, shape (n_components, n_features)</em>) – Initial guesses for solving X_atac ~= WH_atac</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>C. Boutsidis, E. Gallopoulos: SVD based initialization: A head start for
nonnegative matrix factorization - Pattern Recognition, 2008,
<a class="reference external" href="http://tinyurl.com/nndsvd">http://tinyurl.com/nndsvd</a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="nmf_models.nmf_models_mod_updates.intNMF.fit">
<span class="sig-name descname"><span class="pre">fit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rna_mat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">atac_mat</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nmf_models/nmf_models_mod_updates.html#intNMF.fit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nmf_models.nmf_models_mod_updates.intNMF.fit" title="Permalink to this definition"></a></dt>
<dd><p>Optimise NMF.</p>
<p>Uses accelerated Hierarchical alternating least squares algorithm proposeed here, but modified to
joint factorise two matrices. <a class="reference external" href="https://arxiv.org/pdf/1107.5194.pdf">https://arxiv.org/pdf/1107.5194.pdf</a>. Only required arguments are the matrices to use for factorisation.
GEX and ATAC matrices are expected in cell by feature format. Matrices should be scipy sparse matrices.
min ||X_rna - (theta . phi_rna)||_2 and min ||X_atac - (theta . phi_atac)||_2 s.t. theta, phi_rna, phi_atac &gt; 0. So that theta hold the latent topic scores for a cell. And phi
the allows recontruction of X</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rna_mat</strong> (<em>scipy sparse matrix</em><em> of </em><em>single cell gene expression</em>) – </p></li>
<li><p><strong>atac_mat</strong> (<em>scipy sparse matrix</em><em> of </em><em>single cell gene expression</em>) – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nmf_models.nmf_models_mod_updates.log_tf_idf">
<span class="sig-prename descclassname"><span class="pre">nmf_models.nmf_models_mod_updates.</span></span><span class="sig-name descname"><span class="pre">log_tf_idf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mat_in</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nmf_models/nmf_models_mod_updates.html#log_tf_idf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nmf_models.nmf_models_mod_updates.log_tf_idf" title="Permalink to this definition"></a></dt>
<dd><p>Return a TF-IDF transformed matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mat_in</strong> (<a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html">csr matrix type</a>) – Positional (required) matrix to be TF-IDF transformed. Should be cell x feature.</p></li>
<li><p><strong>scale</strong> (<em>int</em>) – scale values by this number (optional)</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>nmf_models.InvalidMatType</strong> – If the mat type is invalid:</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>TF-IDF transformed matrix</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>csr matrix</p>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="nmf_models.html" class="btn btn-neutral float-left" title="nmf_models package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="nmf_models.plotting.html" class="btn btn-neutral float-right" title="nmf_models.plotting module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, william morgans.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>